### PR TITLE
Yago/step12.2

### PR DESCRIPTION
**Pull request by YagoAccenture**
Deletion of entity in domain layer, and created in the presentation layer

## Release Notes

- **New Features**
  - Introduced a new use case for fetching episode details and associated TMDB information.
  - Enhanced `DetailsActivity` to display episode ratings and images.
  - Added new UI components for episode rating and imagery in the layout.
  - New data classes created to represent episode details and season/episode relationships.

- **Bug Fixes**
  - Corrected typographical errors in class and method names related to TMDB functionality.

- **Improvements**
  - Updated data handling for episode details to include TMDB data, improving overall functionality and user experience.

- **Style**
  - Improved layout structure and visual alignment of UI components in the details screen.
  - Added a new string resource for displaying character ratings.

### CODE REVIEW

**Review by PabloGrelaM — CHANGES_REQUESTED** [review]
Review GetEpisodeAndDetailsUseCase

File: app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import com.mstudio.superheroarch.presentation.mapToEpisodeDetailsViewEntity
+import com.mstudio.superheroarch.repository.RickAndMortyRepository
+import com.mstudio.superheroarch.repository.TmdbRepository
+
+class GetEpisodeAndDetailsUseCase(
+    private val repository: RickAndMortyRepository,
+    private val tmdbRepository: TmdbRepository
+) {
+    suspend operator fun invoke(episodeUrl: String): Result<EpisodeDetailsViewEntity> {
+        val episodeResult = repository.fetchEpisodeDetails(episodeUrl)
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt]: this repo shouldn't return a Result<T> it should return a T

File: app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                )
+                return Result.success(episodeDetailsViewEntity)
+            } else {
+                return Result.failure(Exception("TMDB episode info is null"))
+            }
+        }
+        return Result.failure(Exception("Failed to load episode details"))
+    }
+
+    private fun extractSeasonAndEpisode(episodeCode: String?): SeasonAndEpisode {
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt]: This should return a nullable object, if the parse fails, a null should be returned.
Also change the substring for a REGEX.

File: app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            val episodeEntity = episodeResult.getOrThrow()
+
+            val seasonAndEpisode = extractSeasonAndEpisode(episodeEntity.episode)
+            val seasonNumber = seasonAndEpisode.seasonNumber ?: 0
+            val episodeNumber = seasonAndEpisode.episodeNumber ?: 0
+
+            val tmdbResult = tmdbRepository.getEpisodeTMDBInfo(seasonNumber, episodeNumber)
+
+            if (tmdbResult != null) {
+                val episodeDetailsViewEntity = mapToEpisodeDetailsViewEntity(
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt]: change this map to be consistent with the rest of the project

File: app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt [hunk 4]
```diff
@@ -23,18 +23,25 @@ class DetailsViewModel(
 
     private suspend fun fetchFirstEpisodeDetails(episodeUrl: String) {
         try {
-            val episodeResult = getEpisodeDetailsUseCase(episodeUrl)
-            episodeResult.getOrNull()?.let {
-                view.displayFirstEpisodeDetails(it)
+            val result = getEpisodeAndDetailsUseCase(episodeUrl)
+            if (result.isSuccess) {
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt]: unsae result.onSuccess{}.onFailure{}

File: app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt [hunk 5]
```diff
@@ -23,18 +23,25 @@ class DetailsViewModel(
...
+            view.showError("Failed to load episode details: ${e.message}")
         }
     }
+
 }
 
 interface DetailsViewTranslator {
     fun displayCharacterDetails(character: CharacterEntity)
-    fun displayFirstEpisodeDetails(episode: EpisodeEntity)
+    fun displayFirstEpisodeDetails(episodeDetailsViewEntity: EpisodeDetailsViewEntity)
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt]: don't use two different methods with the same param and being called in the same method.
Split them to handle the option of one backend failing and the other working or make the assumption that both are going to work.

File: app/src/main/java/com/mstudio/superheroarch/presentation/SeasonAndEpisode.kt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.mstudio.superheroarch.presentation
+
+data class SeasonAndEpisode(
+    val seasonNumber: Int?,
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/presentation/SeasonAndEpisode.kt]: those vals are not nullable

File: app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt [hunk 7]
```diff
@@ -66,6 +77,23 @@ class DetailsActivity : AppCompatActivity(), DetailsViewTranslator {
...
 
+    override fun displayEpisodeRatingAndImage(episodeDetailsViewEntity: EpisodeDetailsViewEntity) {
+        val ratingTextView: TextView = findViewById(R.id.episode_rating)
+        val episodeImageView: ImageView = findViewById(R.id.episode_image)
+        Log.d("DetailsActivity", "Displaying episode rating: ${episodeDetailsViewEntity.rating}, imageUrl: ${episodeDetailsViewEntity.imageUrl}")
+
+        ratingTextView.text = "Rating: ${episodeDetailsViewEntity.rating}"
+
+        episodeDetailsViewEntity.imageUrl.let {
+            Picasso.get().load(it)
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt]: this let is not needed.

File: app/src/main/res/layout/details_screen.xml [hunk 8]
```diff
@@ -203,7 +206,7 @@
                 android:id="@+id/first_episode"
                 android:layout_width="wrap_content"
                 android:layout_height="wrap_content"
-                android:layout_marginTop="8dp"
+                android:layout_marginTop="0dp"
```
Comment (PabloGrelaM) [inline app/src/main/res/layout/details_screen.xml]: if margin is 0, don't set any margin.

**Review by PabloGrelaM — CHANGES_REQUESTED** [review]
After this PR, create another one creating Models for each layer.

File: app/src/main/res/layout/details_screen.xml (hunk 8 above)
Comment (YagoAccenture) [reply app/src/main/res/layout/details_screen.xml]: Solved in next commit.

File: app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt (hunk 7 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt]: Solved in next commit.

File: app/src/main/java/com/mstudio/superheroarch/presentation/SeasonAndEpisode.kt (hunk 6 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/presentation/SeasonAndEpisode.kt]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt (hunk 5 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt (hunk 2 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt (hunk 4 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt]: Solved in next commit¿?

File: app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt (hunk 3 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/domain/GetEpisodeAndDetailsUseCase.kt]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt (hunk 4 above)
Comment (PabloGrelaM) [reply app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt]: nope

File: app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package com.mstudio.superheroarch.presentation
+
+data class EpisodeDetailsViewEntity(
+    val air_date: String,
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt]: air_date is not a valid naming

File: app/src/main/java/com/mstudio/superheroarch/repository/RickAndMortyRepository.kt [hunk 10]
```diff
@@ -29,12 +29,11 @@ class RickAndMortyRepository(
...
-    suspend fun fetchEpisodeDetails(episodeUrl: String): Result<EpisodeEntity>{
+    suspend fun fetchEpisodeDetails(episodeUrl: String): EpisodeEntity {
         return try {
-            val response = remoteDataSource.getEpisode(episodeUrl)
-            Result.success(response)
-        }catch (e: IOException){
-            Result.failure(e)
+            remoteDataSource.getEpisode(episodeUrl)
+        } catch (e: IOException) {
+            throw e
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/repository/RickAndMortyRepository.kt]: catch + throw means nothing. You are catching the exception and throwing it again, this is useless. Remove the try catch.

File: app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt [hunk 11]
```diff
@@ -1,6 +1,7 @@
 package com.mstudio.superheroarch.ui.activity
 
 import android.os.Bundle
+import android.util.Log
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt]: unused import.

**Review by PabloGrelaM — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt (hunk 11 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/repository/RickAndMortyRepository.kt (hunk 10 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/repository/RickAndMortyRepository.kt]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt (hunk 9 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt (hunk 4 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt]: I think that now is solved in the next commit.

File: app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt [hunk 12]
```diff
@@ -23,18 +24,23 @@ class DetailsViewModel(
 
     private suspend fun fetchFirstEpisodeDetails(episodeUrl: String) {
         try {
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt]: You are doing the try catch in the usecase, that catch returns a Result.failure, so you don't need this try catch.

File: app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+data class EpisodeDetailsViewEntity(
+    val airDate: String,
+    val episode: String,
+    val rating: Double,
+    val imageUrl: String
+)
+
+fun EpisodeDetailsViewEntity.mapToViewEntity(
+    air_date: String,
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt]: air_date is still not a valid naming.

File: app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt [hunk 14]
```diff
@@ -76,9 +86,24 @@ class DetailsActivity : AppCompatActivity(), DetailsViewTranslator {
...
     }
 
-    override fun displayFirstEpisodeDetails(episode: EpisodeEntity) {
-        firstEpisodeTextView?.text = episode.episode
-        firstEpisodeDateTextView?.text = episode.air_date
+    override fun displayEpisodeDetails(
+        episodeEntity: EpisodeEntity?,
+        episodeDetailsViewEntity: EpisodeDetailsViewEntity?
+    ) {
+        if (episodeDetailsViewEntity != null) {
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/ui/activity/DetailsActivity.kt]: this if should be in the viewmodel.

**Review by PabloGrelaM — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt (hunk 13 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/presentation/EpisodeDetailsViewEntity.kt]: Solved in next comit

File: app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt (hunk 12 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/presentation/DetailsViewModel.kt]: Solved in next commit

**Review by PabloGrelaM — APPROVED** [review]

**Pull request merged**